admob/src/android/banner_view_internal_android.cc

```diff
@@ -66,37 +99,131 @@ void DestroyOnDeleteCallback(const Future<void>& result, void* sem_data) {
+  jobject banner_view_helper;
+  FutureCallbackData* callback_data;
+};
+
+// This function is run on the main thread and is called in the
+// BannerViewInternalAndroid::Initialize() method.
+void InitializeBannerViewOnMainThread(void* data) {
+  BannerViewInternalInitializeData* call_data =
+      reinterpret_cast<BannerViewInternalInitializeData*>(data);
+  JNIEnv* env = GetJNI();
+  FIREBASE_ASSERT(env != nullptr);
+
+  jstring ad_unit_id_str = env->NewStringUTF(call_data->ad_unit_id.c_str());
+  env->CallVoidMethod(call_data->ad_view,
+                      ad_view::GetMethodId(ad_view::kSetAdUnitId),
+                      ad_unit_id_str);
+  env->DeleteLocalRef(ad_unit_id_str);
+
+  bool jni_exception = util::CheckAndClearJniExceptions(env);
+  FIREBASE_ASSERT(!jni_exception);
+
+  jobject ad_size = env->NewObject(
+      ad_size::GetClass(), ad_size::GetMethodId(ad_size::kConstructor),
+      call_data->ad_size.width, call_data->ad_size.height);
+
+  jni_exception = util::CheckAndClearJniExceptions(env);
+  FIREBASE_ASSERT(!jni_exception);
+
+  env->CallVoidMethod(call_data->ad_view,
+                      ad_view::GetMethodId(ad_view::kSetAdSize), ad_size);
+
+  jni_exception = util::CheckAndClearJniExceptions(env);
+  FIREBASE_ASSERT(!jni_exception);
+
+  env->CallVoidMethod(
+      call_data->banner_view_helper,
+      banner_view_helper::GetMethodId(banner_view_helper::kInitialize),
+      call_data->activity_global);
+
+  jni_exception = util::CheckAndClearJniExceptions(env);
+  FIREBASE_ASSERT(!jni_exception);
+
+  jobject ad_listener = nullptr;
+  ad_listener =
+      env->NewObject(banner_view_helper_ad_view_listener::GetClass(),
+                     banner_view_helper_ad_view_listener::GetMethodId(
+                         banner_view_helper_ad_view_listener::kConstructor),
+                     call_data->banner_view_helper);
+
+  jni_exception = util::CheckAndClearJniExceptions(env);
+  FIREBASE_ASSERT(!jni_exception);
+
+  env->CallVoidMethod(call_data->ad_view,
+                      ad_view::GetMethodId(ad_view::kSetAdListener),
+                      ad_listener);
+
+  jni_exception = util::CheckAndClearJniExceptions(env);
+  FIREBASE_ASSERT(!jni_exception);
+
+  env->DeleteLocalRef(ad_listener);
+
+  CompleteFuture(kAdMobErrorNone, "", call_data->callback_data->future_handle,
+                 call_data->callback_data->future_data);
+
+  delete call_data;
+}
+
 Future<void> BannerViewInternalAndroid::Initialize(AdParent parent,
                                                    const char* ad_unit_id,
                                                    AdSize size) {
   FutureCallbackData* callback_data =
       CreateFutureCallbackData(&future_data_, kBannerViewFnInitialize);
 
+  if (initialized_) {
+    CompleteFuture(kAdMobErrorAlreadyInitialized, "Ad is already initialized.",
+                   callback_data->future_handle, callback_data->future_data);
+    return GetLastResult(kBannerViewFnInitialize);
```

Comment:
Note that this pattern introduces a race condition that we should try to avoid in new code. See b/110082243 and cl/217546936 for details.

Reply:
I believe I fixed this. Please take a look.